File: backend/src/main/java/com/just/donate/utils/Reservable.java
package com.just.donate.utils;

import java.util.Optional;

public class Reservable<R extends Splittable<R, S>, S, C> {

    private final R value;
    private Optional<C> context;

    public Reservable(R value) {
        this.value = value;
        this.context = Optional.empty();
    }

    public boolean isReserved() {
        return context.isPresent();
    }
    
    public boolean isReservedBy(C context) {
        return this.context.map(c -> c.equals(context)).orElse(false);
    }

    public void reserve(C context) {
        this.context = Optional.of(context);
    }

    public void release() {
        this.context = Optional.empty();
    }

    public R getValue() {
        return value;
    }

    public Optional<C> getContext() {
        return context;
    }

    @Override
    public int hashCode() {
        return super.hashCode();
    }

    @Override
    public boolean equals(Object obj) {
        return super.equals(obj);
    }
    
    @Override
    public String toString() {
        return context
                .map(c -> String.format("%s(%s)", c, value))
                .orElseGet(value::toString);
    }
}

File: backend/src/main/scala/com/just/donate/utils/Reservable.scala
package com.just.donate.utils

case class Reservable[R <: Splittable[R, S], S, C](value: R, context: Option[C] = None):

  def isReserved: Boolean = context.isDefined

  def isReservedBy(ctx: C): Boolean = context.exists(c => c.equals(ctx))

  def reserve(ctx: C): Reservable[R, S, C] = Reservable(value, Option(ctx))

  def release(): Reservable[R, S, C] = Reservable(value, None)

  override def toString: String =
    if context.isDefined then String.format("%s(%s)", context.get, value)
    else value.toString
